### PR TITLE
MM-20904 Add E2E tests for custom categories

### PR DESCRIPTION
**Pull request by sudheerDev**
#### Summary
 * Tests for create, delete and rename categories

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-20904

### CODE REVIEW

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        cy.apiUpdateConfig({
+            ServiceSettings: {
+                ExperimentalChannelSidebarOrganization: 'default_on',
+            },
+        });
+
+        // # Login as test user and visit town-square
+        cy.apiInitSetup({loginAfter: true}).then(({team}) => {
+            cy.visit(`/${team.name}/channels/town-square`);
+        });
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:26]: Actually, there's a new `uiCloseWhatsNewModal` added by https://github.com/mattermost/mattermost-webapp/pull/5984

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        const categoryName = createCategoryFromSidebarMenu();
+
+        // # create a category from cateory menu
+        cy.get(`button[aria-label='${categoryName}']`).parents('.SidebarChannelGroup').within(() => {
+            cy.get('.SidebarMenu').invoke('show').get('.SidebarMenu_menuButton').click();
+            cy.get('.icon-folder-plus-outline').parents('button').click();
+        });
+
+        const newCategoryName = `category-${getRandomId()}`;
+        cy.get('#editCategoryModal input').type(newCategoryName).type('{enter}');
```
[the remark this answers is not in the document]
Comment (sudheerDev) [reply e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:46]: wierd, let me check that

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+// ***************************************************************
+
+// Stage: @prod
+// Group: @channel_sidebar
+
+import {getRandomId} from '../../utils';
+
+describe('Channel sidebar', () => {
+    before(() => {
+        cy.apiUpdateConfig({
```
Comment (Willyfrog) [inline e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:17]:
```suggestion
    before(() => {
        // # Enable channel sidebar organization
        cy.apiUpdateConfig({
```

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        const categoryName = createCategoryFromSidebarMenu();
+
+        // * check if the category exists
+        cy.get(`button[aria-label='${categoryName}']`).should('contain', categoryName);
+    });
+
+    it('should create a new category from cateory menu', () => {
+        const categoryName = createCategoryFromSidebarMenu();
+
+        // # create a category from cateory menu
```
Comment (Willyfrog) [inline e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:39]:
```suggestion
        // # create a category from category menu
```

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sudheerDev) [reply e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:27]: Done

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sudheerDev) [reply e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:50]: ya, changes to master made this invalid, Added ID to the modal, it already has a generic modal id such as labels might as well add one to the modal

**Comment by hmhealey** [comment]
Does this need QA review? I don't know what the process usually is for those

**Comment by sudheerDev** [comment]
going to add TM4J id's for tests

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        // # Login as test user and visit town-square
+        cy.apiInitSetup({loginAfter: true}).then(({team}) => {
+            cy.visit(`/${team.name}/channels/town-square`);
+        });
+
+        // # Close "What's new" modal
+        cy.uiCloseWhatsNewModal();
+    });
+
+    it('should create a new category from sidebar menu', () => {
```
Comment (jgilliam17) [inline e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:33]:
```suggestion
    it('MM-T3161_1 should create a new category from sidebar menu', () => {
```

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        const categoryName = createCategoryFromSidebarMenu();
+
+        // * check if the category exists
+        cy.get(`button[aria-label='${categoryName}']`).should('contain', categoryName);
+    });
+
+    it('should create a new category from category menu', () => {
```
Comment (jgilliam17) [inline e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:40]:
```suggestion
    it('MM-T3161_2 should create a new category from category menu', () => {
```

File: e2e/cypress/integration/channel_sidebar/custom_categories_spec.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        });
+
+        const newCategoryName = `category-${getRandomId()}`;
+        cy.get('#editCategoryModal input').type(newCategoryName).type('{enter}');
+
+        // * check if the newly created category exists
+        cy.get(`button[aria-label='${newCategoryName}']`).should('contain', newCategoryName);
+    });
+
+    it('move an existing channel to a new category', () => {
```
Comment (jgilliam17) [inline e2e/cypress/integration/channel_sidebar/custom_categories_spec.js:56]:
```suggestion
    it('MM-T3161_3 move an existing channel to a new category', () => {
```

**Review by jgilliam17 — CHANGES_REQUESTED** [review]

**Review by jgilliam17 — APPROVED** [review]
Thank you @sudheerDev 
Looks good to merge.
- Updated related TM4J test case steps, details and links.

**Pull request merged**